pallets/stateful-storage/src/lib.rs

```diff
@@ -260,17 +263,94 @@ pub mod pallet {
 
 		#[pallet::call_index(1)]
 		#[pallet::weight(0)]
-		pub fn upsert_page(_origin: OriginFor<T>, payload: Vec<u8>) -> DispatchResult {
+		pub fn upsert_page(
+			origin: OriginFor<T>,
+			state_owner_msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			page_id: PageId,
+			payload: Vec<u8>,
+		) -> DispatchResult {
 			ensure!(
 				payload.len() as u32 <= T::MaxPaginatedPageSizeBytes::get(),
 				Error::<T>::PageExceedsMaxPageSizeBytes
 			);
+			ensure!(
+				page_id as u32 <= T::MaxPaginatedPageId::get(),
+				Error::<T>::PageIdExceedsMaxAllowed
+			);
+			let provider_key = ensure_signed(origin)?;
```

Comment:
let provider_key = ensure_signed(origin)?;
I would move this check to the top since it checks that this transaction's origin is signed.

Reply:
I've refactored this into a `check_origin` method that also gets the `MessageSourceId`. But I'm not calling it first thing because that requires a DB read; I figured better to check the payload first that doesn't require any DB reads.

Reply:
Take a look at the latest commit; refactored function `check_origin` does both signature verification AND looks up MSA (incurs DB read). If we really want signature verification to come first ALWAYS, I can rework it a little.

Reply:
OK, refactored once more. Should be good now.